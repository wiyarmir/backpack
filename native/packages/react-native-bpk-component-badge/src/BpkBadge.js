/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */

import React from 'react';
import { View, ViewPropTypes } from 'react-native';
import PropTypes from 'prop-types';
import BpkText from 'react-native-bpk-component-text';
import { getStyle } from './styles';

export const BADGE_TYPES = {
  success: 'success',
  warning: 'warning',
  destructive: 'destructive',
  light: 'light',
  inverse: 'inverse',
  outline: 'outline',
};

export const BADGE_DOCKED_TYPES = {
  left: 'left',
  right: 'right',
};

export type Props = {
  message: string,
  docked: $Keys<typeof BADGE_DOCKED_TYPES>,
  type: $Keys<typeof BADGE_TYPES>,
  style: ?(Object | Array<Object>),
};

const BpkBadge = (props: Props) => {
  const { message, docked, type, style: userStyle } = props;
  const {
    border: borderStyle,
    view: innerStyle,
    text: textStyle,
    outline: outlineStyle,
  } = getStyle(docked, type);

  const innerViewStyle =
    type === BADGE_TYPES.outline
      ? [innerStyle, outlineStyle, borderStyle]
      : [innerStyle, borderStyle];

  return (
    <View style={userStyle}>
      <View style={innerViewStyle}>
        <BpkText style={textStyle} textStyle="xs">
          {message}
        </BpkText>
      </View>
    </View>
  );
};

const propTypes = {
  message: PropTypes.string.isRequired,
  docked: PropTypes.oneOf(Object.keys(BADGE_DOCKED_TYPES)),
  style: ViewPropTypes.style,
  type: PropTypes.oneOf(Object.keys(BADGE_TYPES)),
};

BpkBadge.propTypes = propTypes;

BpkBadge.defaultProps = {
  docked: null,
  type: BADGE_TYPES.warning,
  style: null,
};

export default BpkBadge;
export { propTypes };
