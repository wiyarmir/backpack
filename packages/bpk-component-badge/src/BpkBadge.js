/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import PropTypes from 'prop-types';
import React from 'react';
import { cssModules } from 'bpk-react-utils';

import STYLES from './bpk-badge.scss';

export const BADGE_TYPES = {
  WARNING: 'warning',
  SUCCESS: 'success',
  DESTRUCTIVE: 'destructive',
  LIGHT: 'light',
  INVERSE: 'inverse',
  OUTLINE: 'outline',
};

const getClassName = cssModules(STYLES);

const badgeTypeClassNames = {
  [BADGE_TYPES.WARNING]: null,
  [BADGE_TYPES.SUCCESS]: getClassName('bpk-badge--success'),
  [BADGE_TYPES.DESTRUCTIVE]: getClassName('bpk-badge--destructive'),
  [BADGE_TYPES.LIGHT]: getClassName('bpk-badge--light'),
  [BADGE_TYPES.INVERSE]: getClassName('bpk-badge--inverse'),
  [BADGE_TYPES.OUTLINE]: getClassName('bpk-badge--outline'),
};

const BpkBadge = props => {
  const { type, docked, centered, className, ...rest } = props;
  const classNames = [getClassName('bpk-badge'), badgeTypeClassNames[type]];

  if (docked === 'right') {
    classNames.push(getClassName('bpk-badge--docked-right'));
  }
  if (docked === 'left') {
    classNames.push(getClassName('bpk-badge--docked-left'));
  }
  if (centered) {
    classNames.push(getClassName('bpk-badge--centered'));
  }
  if (className) {
    classNames.push(className);
  }

  return <span className={classNames.join(' ')} {...rest} />;
};

BpkBadge.propTypes = {
  type: PropTypes.oneOf([
    BADGE_TYPES.WARNING,
    BADGE_TYPES.SUCCESS,
    BADGE_TYPES.DESTRUCTIVE,
    BADGE_TYPES.LIGHT,
    BADGE_TYPES.INVERSE,
    BADGE_TYPES.OUTLINE,
  ]),
  docked: PropTypes.oneOf(['right', 'left', null]),
  centered: PropTypes.bool,
  className: PropTypes.string,
};

BpkBadge.defaultProps = {
  type: BADGE_TYPES.WARNING,
  docked: null,
  centered: false,
  className: null,
};

export default BpkBadge;
