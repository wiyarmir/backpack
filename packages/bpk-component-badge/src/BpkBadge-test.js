/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import BpkBadge, { BADGE_TYPES } from './BpkBadge';

describe('BpkBadge', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<BpkBadge>Promociando</BpkBadge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a "centered"', () => {
    const tree = renderer
      .create(<BpkBadge centered>Promociando</BpkBadge>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a "docked" attribute value equal to "right"', () => {
    const tree = renderer
      .create(<BpkBadge docked="right">Promociando</BpkBadge>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a "docked" attribute value equal to "left"', () => {
    const tree = renderer
      .create(<BpkBadge docked="left">Promociando</BpkBadge>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when success', () => {
    const tree = renderer
      .create(<BpkBadge type={BADGE_TYPES.SUCCESS}>Promociando</BpkBadge>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when destructive', () => {
    const tree = renderer
      .create(<BpkBadge type={BADGE_TYPES.DESTRUCTIVE}>Promociando</BpkBadge>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when light', () => {
    const tree = renderer
      .create(<BpkBadge type={BADGE_TYPES.LIGHT}>Promociando</BpkBadge>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when inverse', () => {
    const tree = renderer
      .create(<BpkBadge type={BADGE_TYPES.INVERSE}>Promociando</BpkBadge>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when outline', () => {
    const tree = renderer
      .create(<BpkBadge type={BADGE_TYPES.OUTLINE}>Promociando</BpkBadge>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
