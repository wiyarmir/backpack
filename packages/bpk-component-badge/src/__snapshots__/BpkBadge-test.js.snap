// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkBadge should render correctly 1`] = `
<span
  className="bpk-badge "
>
  Promociando
</span>
`;

exports[`BpkBadge should render correctly when light 1`] = `
<span
  className="bpk-badge bpk-badge--light"
>
  Promociando
</span>
`;

exports[`BpkBadge should render correctly when success 1`] = `
<span
  className="bpk-badge bpk-badge--success"
>
  Promociando
</span>
`;

exports[`BpkBadge should render correctly when outline 1`] = `
<span
  className="bpk-badge bpk-badge--outline"
>
  Promociando
</span>
`;

exports[`BpkBadge should render correctly when destructive 1`] = `
<span
  className="bpk-badge bpk-badge--destructive"
>
  Promociando
</span>
`;

exports[`BpkBadge should render correctly when inverse 1`] = `
<span
  className="bpk-badge bpk-badge--inverse"
>
  Promociando
</span>
`;

exports[`BpkBadge should render correctly with a "centered" 1`] = `
<span
  className="bpk-badge  bpk-badge--centered"
>
  Promociando
</span>
`;

exports[`BpkBadge should render correctly with a "docked" attribute value equal to "left" 1`] = `
<span
  className="bpk-badge  bpk-badge--docked-left"
>
  Promociando
</span>
`;

exports[`BpkBadge should render correctly with a "docked" attribute value equal to "right" 1`] = `
<span
  className="bpk-badge  bpk-badge--docked-right"
>
  Promociando
</span>
`;
